### ISSUE TITLE
High CPU usage with 1.0.781-prerelease

### ISSUE THREAD

**Issue by AbeniMatteo** [MicrosoftEdge/WebView2Feedback#878]
**Description**
High CPU usage while doing nothing. This problem is only present from `1.0.781-prerelease`.
`1.0.721-prerelease` and `1.0.705.50` are ok.

**Version**
SDK: 1.0.781-prerelease
Runtime: 89.0.774.8 (Official build) dev (64-bit)
Framework: .NET 5.0.102 + WPF
OS: Win10 10.0.19042 Build 19042

**Repro Steps**
Run the following code:

Program.cs
```cs
namespace Repro
{
    using System;
    using System.Windows;
    using Microsoft.Web.WebView2.Wpf;

    static class Program
    {
        [STAThread]
        static void Main()
        {
            // Init Application.Current
            new Application();
            var w = new Window();
            w.Content = new WebView2();
            w.ShowDialog();
        }
    }
}
```

Repro.csproj
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>WinExe</OutputType>
    <TargetFramework>net5.0-windows</TargetFramework>
    <UseWPF>true</UseWPF>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.Web.WebView2" Version="1.0.781-prerelease" />
  </ItemGroup>
</Project>
```

[AB#31577027](https://microsoft.visualstudio.com/90b2a23c-cab8-4e7c-90e7-a977f32c1f5d/_workitems/edit/31577027)

**Comment by albahari** [comment]
I've reproduced this problem with both the WPF and WinForms controls.

It's caused by a continuous stream of WM_PAINT messages being generated, keeping the UI thread close to 100% busy. This also saturates the message loop and sometimes prevents other controls from receiving a WM_PAINT. You can verify this as follows:

    public class WebView2Ex : WebView2
    {
        protected override IntPtr WndProc (IntPtr hwnd, int msg, IntPtr wParam, IntPtr lParam, ref bool handled)
        {
            Debug.WriteLine (msg);
            return base.WndProc (hwnd, msg, wParam, lParam, ref handled);
        }
    }

This needs to be fixed ASAP, because the control is close to unusable right now.

**Comment by tildebox** [comment]
15% CPU usage. I had to downgrade to 1.0.705.50.

**Comment by champnic** [comment]
Thanks for finding this - I've filed a bug on our backlog and we're taking a look right away.

**Comment by johna-ms** [comment]
Hello all! We are working on getting another release turned around as soon as possible with the fix for this. In the mean time, you can try this workaround:

- Extend the WebView control with your own control
- Override the WndProc method
- Check for the WM_PAINT window message (0x000F), call BeginPaint, EndPaint, and return IntPtr.Zero
  - Begin/EndPaint are Win32 APIs and you'll need to DllImport them. Here's the code we used:
  -      `[StructLayout(LayoutKind.Sequential)]
            public struct Rect
            {
                public int left;
                public int top;
                public int right;
                public int bottom;
            }

            [StructLayout(LayoutKind.Sequential)]
            public struct PaintStruct
            {
                public IntPtr hdc;
                public bool fErase;
                public Rect rcPaint;
                public bool fRestore;
                public bool fIncUpdate;
                [MarshalAs(UnmanagedType.ByValArray, SizeConst=32)] public byte[] rgbReserved;
            }

            [DllImport("user32.dll", SetLastError = true)]
            public static extern IntPtr BeginPaint(IntPtr hwnd, out PaintStruct lpPaint);

            [DllImport("user32.dll", SetLastError = true)]
            public static extern bool EndPaint(IntPtr hwnd, ref PaintStruct lpPaint);`

**Comment by ukandrewc** [comment]
@johna-ms To save me having to work it out, what are the values for PaintStruct, please.

Thanks

**Comment by ukandrewc** [comment]
This sort of works, but still got the issues with z-order. Any WebView2 that paints, comes to the front.

**Comment by johna-ms** [comment]
@ukandrewc does your code in WndProc look like this?

        `protected override void WndProc(ref Message m)
        {
            switch (unchecked((NativeMethods.WM)m.Msg))
            {
                case NativeMethods.WM.PAINT:
                {
                    NativeMethods.PaintStruct lpPaint;
                    NativeMethods.BeginPaint(m.HWnd, out lpPaint);
                    NativeMethods.EndPaint(m.HWnd, ref lpPaint);
                    m.Result = IntPtr.Zero;
                    return;
                }
            }
            base.WndProc(ref m);
        }`

Re: your last message, you don't need to fill out PaintStruct at all

**Comment by ukandrewc** [comment]
@johna-ms Yes, it does (I work in VB, but it's the same), for anyone else in VB:

```
Imports System.Runtime.InteropServices
Public Class WVBrowser : Inherits WebView2

	<StructLayout(LayoutKind.Sequential)>
	Private Structure Rect
		Public left As Integer
		Public top As Integer
		Public right As Integer
		Public bottom As Integer
	End Structure

	<StructLayout(LayoutKind.Sequential)>
	Private Structure PaintStruct
		Public hdc As IntPtr
		Public fErase As Boolean
		Public rcPaint As Rect
		Public fRestore As Boolean
		Public fIncUpdate As Boolean
		<MarshalAs(UnmanagedType.ByValArray, SizeConst:=32)>
		Public rgbReserved As Byte()
	End Structure

	<DllImport("user32.dll", SetLastError:=True)>
	Private Shared Function BeginPaint(ByVal hwnd As IntPtr, <Out> ByRef lpPaint As PaintStruct) As IntPtr
	End Function

	<DllImport("user32.dll", SetLastError:=True)>
	Private Shared Function EndPaint(ByVal hwnd As IntPtr, ByRef lpPaint As PaintStruct) As Boolean
	End Function

	Protected Overrides Sub WndProc(ByRef m As Message)
		If m.Msg = &HF Then
			Dim PS As New PaintStruct
			BeginPaint(m.HWnd, PS)
			EndPaint(m.HWnd, PS)
			m.Result = IntPtr.Zero
		Else
			MyBase.WndProc(m)
		End If
	End Sub

End Class

```



Thanks, I deleted the question once I'd read your code properly (been a while since I've done any winapi) ;-)

**Comment by johna-ms** [comment]
So this fixes the CPU issue for you but not the related issue about items refreshing on the form?

**Comment by ukandrewc** [comment]
Yes, sorry, just realised this issue is only for High CPU usage, not #875.

It does allow other controls on the form to refresh normally, but I still still seeing transparency issues with Z order (Shown in the video above)

I also had to add Web2.Refresh() after Web2.BringToFront()

**Comment by ukandrewc** [comment]
It seems that whichever WebView2 paints last gets to be visible, but this doesn't affect main scrollbars.

When Web2 is at the front of Z order, and Web1 is refreshed, it becomes visible. Scrolling Web2, shows it's content again (until Web1 content changes).

**Comment by johna-ms** [comment]
Ok that's definitely weird. We're tracking it on our end as a separate issue. Thanks!

**Comment by ukandrewc** [comment]
In the video Web2 is at the front (HTML Table)
When Web1 banner (eBay) refreshes, it is visible.
Scrolling Web2 (scrollbar is Web2's) shows it again (until eBay content changes)

https://user-images.githubusercontent.com/30246320/106948454-a3874680-6723-11eb-95e7-2fe68de9ecd7.mp4

**Comment by johna-ms** [comment]
Tracking this on github here

https://github.com/MicrosoftEdge/WebView2Feedback/issues/897

**Comment by ukandrewc** [comment]
Thanks

**Comment by champnic** [comment]
The high CPU usage should now be fixed in 1.0.790-prerelease SDK package. Thanks!
